internal/kubeutils.go

```diff
@@ -22,33 +22,37 @@ import (
 func UpdateNodeRetry(client corev1client.NodeInterface, lister corev1lister.NodeLister, nodeName string, f func(*corev1.Node)) (*corev1.Node, error) {
 	var node *corev1.Node
 	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
-		n, err := lister.Get(nodeName)
-		if err != nil {
-			return err
-		}
-		oldNode, err := json.Marshal(n)
-		if err != nil {
-			return err
-		}
-
-		nodeClone := n.DeepCopy()
-		f(nodeClone)
-
-		newNode, err := json.Marshal(nodeClone)
-		if err != nil {
-			return err
-		}
-
-		patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldNode, newNode, corev1.Node{})
-		if err != nil {
-			return fmt.Errorf("failed to create patch for node %q: %v", nodeName, err)
-		}
-
-		node, err = client.Patch(context.TODO(), nodeName, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
-		return err
+		var innerErr error
+		node, innerErr = UpdateNode(client, lister, nodeName, f)
+		return innerErr
 	}); err != nil {
 		// may be conflict if max retries were hit
-		return nil, fmt.Errorf("unable to update node %q: %v", node, err)
+		return nil, fmt.Errorf("unable to update node %q: %v", nodeName, err)
 	}
 	return node, nil
 }
+
+// UpdateNode calls f to update a node object in Kubernetes.
+func UpdateNode(client corev1client.NodeInterface, lister corev1lister.NodeLister, nodeName string, f func(*corev1.Node)) (*corev1.Node, error) {
+	n, err := lister.Get(nodeName)
+	if err != nil {
+		return nil, err
+	}
+	oldNodeClone := n.DeepCopy()
+	oldNodeClone.ResourceVersion = ""
```

Comment:
So in the PR you mentioned that:
```
That makes sure that if we try to patch an object using an old resouce version we'll get a conflict error,
we then exit the guard we entered, and retry with a fresher object, all good.
```

I'm a little confused as to how we're actually performing this. By setting the old `ResourceVersion` to empty, the merge should set the ResourceVersion to whatever is in newNode, right? Is that different from before? I don't really see a difference in how ResourceVersion is used I guess is my question.

Reply:
Right, let me explain better the situation here and maybe you can help me word this better Jerry:

Before (just an example, annotations is a map, but I'm doing array for simplicity):

- grab the node object (we'll call it `instance{number: 1, resourceVersion: 42, annotations: ["test1"]}`)
- clone it (now we have `instance{number: 2, resourceVersion: 42, annotations: ["test1"]}`)
- modify the annotations on the clone (now we have `instance{instance{number: 2, resourceVersion: 42, annotations: ["test1", "test2"]}`)
- run the instances through `CreateTwoWayMergePatch`

What `CreateTwoWayMergePatch` does and how Patch works is crucial here. When `CreateTwoWayMergePatch` is run above the following (truncated) `diff` is generated:

```
node
  annotations: ["test1", "test2"]
```

You can notice the above generated a diff, meaning it's giving you just the diff object that we'll use to patch. The result is that Patching with the diff above will only ever change annotations, as that's effectively the only changes `CreateTwoWayMergePatch` can see from the two objects.

The fix I did here forces us to always generate **also** a resourceVersion diff so that it can be noticed at Patch time and the apiserver complains if it's old.

One other thing to note here is that we're not playing with the original `n` node instance, only `oldNodeClone` and `oldNode` to generate the diff (`n` remains untouched)

By setting resourceVersion to empty, we'll go like this:

- grab the node object (we'll call it `instance{number: 1, resourceVersion: 42, annotations: ["test1"]}`)
- make a first clone and empty out the resourceVersion `instance{number: 1, resourceVersion: "", annotations: ["test1"]}`
- clone the original object again (now we have `instance{number: 2, resourceVersion: 42, annotations: ["test1"]}`)
- modify the annotations on this last clone (now we have `instance{instance{number: 2, resourceVersion: 42, annotations: ["test1", "test2"]}`)
- run the instances through `CreateTwoWayMergePatch`

Now, the diff we get is:

```
node
  annotations: ["test1", "test2"]
  resourceVersion: 42
``` 

And we'll patch with the above. If we send `42` but the apiserver already knows about `43` we'll get a conflict

Reply:
Right, let me explain better the situation here and maybe you can help me word this better Jerry:
Before (just an example, annotations is a map, but I'm doing array for simplicity):
- grab the node object (we'll call it `instance{number: 1, resourceVersion: 42, annotations: ["test1"]}`)
- clone it (now we have `instance{number: 2, resourceVersion: 42, annotations: ["test1"]}`)
- modify the annotations on the clone (now we have `instance{instance{number: 2, resourceVersion: 42, annotations: ["test1", "test2"]}`)
- run the instances through `CreateTwoWayMergePatch`
What `CreateTwoWayMergePatch` does and how Patch works is crucial here. When `CreateTwoWayMergePatch` is run above the following (truncated) `diff` is generated:
```
node
  annotations: ["test1", "test2"]
```
You can notice the above generated a diff, meaning it's giving you just the diff object that we'll use to patch. The result is that Patching with the diff above will only ever change annotations, as that's effectively the only changes `CreateTwoWayMergePatch` can see from the two objects.

The fix I did here forces us to always generate **also** a resourceVersion diff so that it can be noticed at Patch time and the apiserver complains if it's old.
One other thing to note here is that we're not playing with the original `n` node instance, only `oldNodeClone` and `oldNode` to generate the diff (`n` remains untouched)
By setting resourceVersion to empty, we'll go like this:
- grab the node object (we'll call it `instance{number: 1, resourceVersion: 42, annotations: ["test1"]}`)
- make a first clone and empty out the resourceVersion `instance{number: 1, resourceVersion: "", annotations: ["test1"]}`
- clone the original object again (now we have `instance{number: 2, resourceVersion: 42, annotations: ["test1"]}`)
- modify the annotations on this last clone (now we have `instance{instance{number: 2, resourceVersion: 42, annotations: ["test1", "test2"]}`)
- run the first old clone and the other clone instances through `CreateTwoWayMergePatch`
Now, the diff we get is:
```
node
  annotations: ["test1", "test2"]
  resourceVersion: 42
``` 

It is ultimately making sure that if we're trying to patch an object, at least the changes we're applying are applied to the very last version of an object in the apiserver, otherwise, we'll retry.

Reply:
Ah, I see, no that makes perfect sense. Awesome explanation!

Maybe for future purposes it'd be nice to add a comment like:

`// Add empty resourceVersion, so CreateTwoWayMergePatch returns a diff with resourceVersion of the node object when comparing.`

Reply:
Adding that, that's really helpful Jerry, thanks for pointing it out!